File: velocity/src/main/java/ru/privatenull/pnauth/velocity/VelocityPlatform.java
package ru.privatenull.pnauth.velocity;

import com.velocitypowered.api.proxy.Player;
import com.velocitypowered.api.proxy.ProxyServer;
import com.velocitypowered.api.scheduler.ScheduledTask;
import ru.privatenull.pnauth.dialog.PlayerDialogs;
import ru.privatenull.pnauth.display.PlayerDisplay;
import ru.privatenull.pnauth.message.MessageFormat;
import ru.privatenull.pnauth.platform.PlatformScheduler;
import ru.privatenull.pnauth.platform.PlatformType;
import ru.privatenull.pnauth.platform.PnPlatform;
import ru.privatenull.pnauth.platform.PnPlayer;
import ru.privatenull.pnauth.platform.TaskHandle;
import ru.privatenull.pnauth.platform.TaskRegistry;
import ru.privatenull.pnauth.platform.DefaultTaskRegistry;

import java.net.InetSocketAddress;
import java.time.Duration;
import java.util.Collection;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

/** Velocity adapter for the public pnAuth platform API. */
public final class VelocityPlatform implements PnPlatform {
    private final Object plugin;
    private final ProxyServer proxy;
    private final PlayerDisplay display;
    private final MessageFormat messageFormat;
    private final PlayerDialogs dialogs;
    private final PlatformScheduler scheduler;
    private final TaskRegistry tasks;

    public VelocityPlatform(Object plugin, ProxyServer proxy, PlayerDisplay display, MessageFormat messageFormat,
                            PlayerDialogs dialogs) {
        this.plugin = plugin;
        this.proxy = proxy;
        this.display = display;
        this.messageFormat = messageFormat;
        this.dialogs = dialogs;
        this.scheduler = new VelocityScheduler();
        this.tasks = new DefaultTaskRegistry(scheduler);
    }

    @Override public PlatformType type() { return PlatformType.VELOCITY; }
    @Override public Optional<PnPlayer> player(UUID uniqueId) { return proxy.getPlayer(uniqueId).map(Wrapper::new); }
    @Override public Optional<PnPlayer> player(String username) { return proxy.getPlayer(username).map(Wrapper::new); }
    @Override public PlatformScheduler scheduler() { return scheduler; }
    @Override public TaskRegistry tasks() { return tasks; }
    @Override public PlayerDialogs dialogs() { return dialogs; }

    @Override
    public Collection<PnPlayer> players() {
        return proxy.getAllPlayers().stream().map(Wrapper::new).map(PnPlayer.class::cast).toList();
    }

    private final class Wrapper implements PnPlayer {
        private final Player delegate;

        private Wrapper(Player delegate) { this.delegate = delegate; }
        @Override public UUID uniqueId() { return delegate.getUniqueId(); }
        @Override public String username() { return delegate.getUsername(); }
        @Override public InetSocketAddress remoteAddress() { return delegate.getRemoteAddress(); }
        @Override public boolean connected() { return delegate.isActive(); }
        @Override public boolean hasPermission(String permission) { return delegate.hasPermission(permission); }
        @Override public PlayerDisplay display() { return display; }
        @Override public PlayerDialogs dialogs() { return dialogs; }
        @Override public PlatformScheduler scheduler() { return scheduler; }

        @Override
        public Optional<String> currentServer() {
            return delegate.getCurrentServer()
                    .map(connection -> connection.getServerInfo().getName());
        }

        @Override public void sendMessage(String message) {
            delegate.sendMessage(VelocityMessages.component(message, messageFormat));
        }
        @Override public void sendMessage(net.kyori.adventure.text.Component message) { delegate.sendMessage(message); }
        @Override public void sendMessages(Iterable<? extends net.kyori.adventure.text.Component> messages) {
            messages.forEach(delegate::sendMessage);
        }

        @Override public void disconnect(String reason) {
            delegate.disconnect(VelocityMessages.component(reason, messageFormat));
        }
    }

    private final class VelocityScheduler implements PlatformScheduler {
        @Override public TaskHandle execute(Runnable task) { return delayed(Duration.ZERO, task); }
        @Override public TaskHandle execute(PnPlayer player, Runnable task) { return execute(task); }
        @Override public TaskHandle delayed(PnPlayer player, Duration delay, Runnable task) { return delayed(delay, task); }
        @Override public TaskHandle repeating(PnPlayer player, Duration initialDelay, Duration interval, Runnable task) {
            return repeating(initialDelay, interval, task);
        }

        @Override
        public TaskHandle delayed(Duration delay, Runnable task) {
            ScheduledTask scheduled = proxy.getScheduler().buildTask(plugin, task)
                    .delay(milliseconds(delay), TimeUnit.MILLISECONDS)
                    .schedule();
            return new ScheduledHandle(scheduled);
        }

        @Override
        public TaskHandle repeating(Duration initialDelay, Duration interval, Runnable task) {
            ScheduledTask scheduled = proxy.getScheduler().buildTask(plugin, task)
                    .delay(milliseconds(initialDelay), TimeUnit.MILLISECONDS)
                    .repeat(Math.max(1L, milliseconds(interval)), TimeUnit.MILLISECONDS)
                    .schedule();
            return new ScheduledHandle(scheduled);
        }

        private long milliseconds(Duration duration) {
            return duration == null ? 0L : Math.max(0L, duration.toMillis());
        }
    }

    private static final class ScheduledHandle implements TaskHandle {
        private final ScheduledTask task;
        private final java.util.concurrent.atomic.AtomicBoolean cancelled = new java.util.concurrent.atomic.AtomicBoolean();

        private ScheduledHandle(ScheduledTask task) { this.task = task; }
        @Override public boolean cancelled() { return cancelled.get(); }
        @Override public void cancel() {
            if (cancelled.compareAndSet(false, true)) task.cancel();
        }
    }
}

File: velocity/src/main/kotlin/ru/privatenull/pnauth/velocity/VelocityPlatform.kt
package ru.privatenull.pnauth.velocity

import com.velocitypowered.api.proxy.ProxyServer
import com.velocitypowered.api.proxy.Player as VPlayer
import ru.privatenull.pnauth.dialog.PlayerDialogs
import ru.privatenull.pnauth.display.PlayerDisplay
import ru.privatenull.pnauth.message.MessageFormat
import ru.privatenull.pnauth.platform.DefaultTaskRegistry
import ru.privatenull.pnauth.platform.Platform
import ru.privatenull.pnauth.platform.PlatformScheduler
import ru.privatenull.pnauth.platform.PlatformType
import ru.privatenull.pnauth.platform.Player
import ru.privatenull.pnauth.platform.TaskHandle
import ru.privatenull.pnauth.platform.TaskRegistry
import java.net.InetSocketAddress
import java.time.Duration
import java.util.Optional
import java.util.UUID
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class VelocityPlatform(
    private val plugin: Any,
    private val proxy: ProxyServer,
    private val display: PlayerDisplay,
    private val messageFormat: MessageFormat,
    private val dialogs: PlayerDialogs
) : Platform {

    private val scheduler: PlatformScheduler = VelocityScheduler(plugin, proxy)
    private val tasks: TaskRegistry = DefaultTaskRegistry(scheduler)

    override fun type(): PlatformType = PlatformType.VELOCITY
    override fun player(uniqueId: UUID): Optional<Player> = wrap(proxy.getPlayer(uniqueId))
    override fun player(username: String): Optional<Player> = wrap(proxy.getPlayer(username))
    override fun scheduler(): PlatformScheduler = scheduler
    override fun tasks(): TaskRegistry = tasks
    override fun dialogs(): PlayerDialogs = dialogs

    override fun players(): Collection<Player> {
        return proxy.allPlayers.map { PPlayer(it) }
    }

    private fun wrap(player: Optional<VPlayer>): Optional<Player> {
        return player.map { PPlayer(it) }
    }

    private inner class PPlayer(private val delegate: VPlayer) : Player {
        override fun uniqueId(): UUID = delegate.uniqueId
        override fun username(): String = delegate.username
        override fun remoteAddress(): InetSocketAddress = delegate.remoteAddress
        override fun connected(): Boolean = delegate.isActive
        override fun hasPermission(permission: String): Boolean = delegate.hasPermission(permission)
        override fun display(): PlayerDisplay = display
        override fun dialogs(): PlayerDialogs = dialogs
        override fun scheduler(): PlatformScheduler = scheduler

        override fun currentServer(): Optional<String> {
            return delegate.currentServer.map { it.serverInfo.name }
        }

        override fun sendMessage(message: String) {
            delegate.sendMessage(VelocityMessages.component(message, messageFormat))
        }

        override fun sendMessage(message: net.kyori.adventure.text.Component) {
            delegate.sendMessage(message)
        }

        override fun sendMessages(messages: Iterable<net.kyori.adventure.text.Component>) {
            messages.forEach { delegate.sendMessage(it) }
        }

        override fun disconnect(reason: String) {
            delegate.disconnect(VelocityMessages.component(reason, messageFormat))
        }
    }

    private class VelocityScheduler(
        private val plugin: Any,
        private val proxy: ProxyServer
    ) : PlatformScheduler {
        override fun execute(task: Runnable): TaskHandle = delayed(Duration.ZERO, task)
        override fun execute(player: Player, task: Runnable): TaskHandle = execute(task)
        override fun delayed(player: Player, delay: Duration, task: Runnable): TaskHandle = delayed(delay, task)
        override fun repeating(player: Player, initialDelay: Duration, interval: Duration, task: Runnable): TaskHandle {
            return repeating(initialDelay, interval, task)
        }

        override fun delayed(delay: Duration, task: Runnable): TaskHandle {
            val scheduled = proxy.scheduler.buildTask(plugin, task)
                .delay(milliseconds(delay), TimeUnit.MILLISECONDS)
                .schedule()
            return ScheduledHandle(scheduled)
        }

        override fun repeating(initialDelay: Duration, interval: Duration, task: Runnable): TaskHandle {
            val scheduled = proxy.scheduler.buildTask(plugin, task)
                .delay(milliseconds(initialDelay), TimeUnit.MILLISECONDS)
                .repeat(Math.max(1L, milliseconds(interval)), TimeUnit.MILLISECONDS)
                .schedule()
            return ScheduledHandle(scheduled)
        }

        companion object {
            private fun milliseconds(duration: Duration?): Long {
                return if (duration == null) 0L else Math.max(0L, duration.toMillis())
            }
        }
    }

    private class ScheduledHandle(private val task: com.velocitypowered.api.scheduler.ScheduledTask) : TaskHandle {
        private val isCancelled = AtomicBoolean()

        override fun cancelled(): Boolean = isCancelled.get()
        override fun cancel() {
            if (isCancelled.compareAndSet(false, true)) task.cancel()
        }
    }
}
